### ISSUE TITLE
Composition of curried composition

### ISSUE THREAD

**Issue by sylee957** [microsoft/pyright#5510]
**Describe the bug**

Type checking fails with error
```python
from collections.abc import Callable
from typing import TypeVar

A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')
D = TypeVar('D')

def compose(g: Callable[[B], C]) -> Callable[[Callable[[A], B]], Callable[[A], C]]:
    return lambda f: lambda a: g(f(a))

def double(g: Callable[[C], D]) -> Callable[[Callable[[A], Callable[[B], C]]], Callable[[A], Callable[[B], D]]]:
    return compose(compose)(compose)(g)
```

```
Expression of type "((Unknown) -> ((((C@double) -> D@double)) -> C@double)) -> ((Unknown) -> ((((C@double) -> D@double)) -> D@double))" cannot be assigned to return type "((A@double) -> ((B@double) -> C@double)) -> ((A@double) -> ((B@double) -> D@double))"
    Type "((Unknown) -> ((((C@double) -> D@double)) -> C@double)) -> ((Unknown) -> ((((C@double) -> D@double)) -> D@double))" cannot be assigned to type "((A@double) -> ((B@double) -> C@double)) -> ((A@double) -> ((B@double) -> D@double))"
      Function return type "(Unknown) -> ((((C@double) -> D@double)) -> D@double)" is incompatible with type "(A@double) -> ((B@double) -> D@double)"
        Type "(Unknown) -> ((((C@double) -> D@double)) -> D@double)" cannot be assigned to type "(A@double) -> ((B@double) -> D@double)"
          Function return type "((C@double) -> D@double) -> D@double" is incompatible with type "(B@double) -> D@double"
            Type "((C@double) -> D@double) -> D@double" cannot be assigned to type "(B@double) -> D@double" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

While using the duplicate function works

```python
def compose(g: Callable[[B], C]) -> Callable[[Callable[[A], B]], Callable[[A], C]]:
    return lambda f: lambda a: g(f(a))

def compose2(g: Callable[[B], C])-> Callable[[Callable[[A], B]], Callable[[A], C]]:
    return compose(g)

def compose3(g: Callable[[B], C])-> Callable[[Callable[[A], B]], Callable[[A], C]]:
    return compose(g)

def double(g: Callable[[C], D]) -> Callable[[Callable[[A], Callable[[B], C]]], Callable[[A], Callable[[B], D]]]:
    return compose(compose2)(compose3)(g)
```

**VS Code extension or command-line**
I'm using pyright `1.1.317`

**Additional context**

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.320, which I just published. It will also be included in a future version of pylance.
